### DISCUSSION TITLE
Why react-select default value provided but still error when required is true

### DISCUSSION THREAD

**Discussion by Cow258** [react-hook-form/react-hook-form#1757] [General] [answered]
Here are demonstration for this issue
https://codesandbox.io/s/unruffled-dhawan-5pr29?file=/src/App.js

Here are two question

1. Why does it show error message when default value provided?
2. After change the selection and submit the form, How can I only get the value of the selection?

I expected to get
```json
{
  "fruit": "apple"
}
```
instead of
```json
{
  "fruit": {
    "value": "apple",
    "label": "Apple"
  }
}
```

Thanks ❤️🙏

**Comment by bluebill1049** [comment, marked as answer]
1) you didn't do anything with `defaultVal`, noting is really selected it just appear to be apple is first in the selection.

2) that's the data return by react-select, to modify that you need to build a wrapper, so it only returns value, not the whole object.

**Reply by Cow258** [reply]
Thank @bluebill1049  for reply.
If I change the `defaultVal` to orange, right side will be orange,\
The `defaultVal` actually work as expected,\
But `react-hook-form` still said it is empty when submit the form.

![螢幕截圖 2020-06-03 上午10 21 40](https://user-images.githubusercontent.com/14261588/83588529-32840380-a584-11ea-901c-c7733e58c118.png)

**Reply by bluebill1049** [reply]
you should set the default value as `Controller` or `useForm` -> `defaultValues` otherwise RHF wouldn't be able to know it's set already. I think the next major version's new `render` props will make this much easier and predictable.

Below is V6 `render` prop
```tsx
<Controller defaultValue="xxx" name="test" render={({ onChange, onBlur, value }) => {
  <Select onChange={onChange} onBlur={onBlur} value={value} />
}} />
```

see `defaultValue` above need to set to something and pass down to React-Select.

**Reply by Cow258** [reply]
After study, I decided to disconnect the linking between `Controller ` and `Select`,\
And call `Controller` `onChange` once when `Select` rendering the first time,\
Also added a `handleChange` to `Select` for convert the value from object to string.

Finally met all my requirements from the beginning,\
But here are big problem!,\
I can't change the value from outside...\
Any solution? 👀

Thanks at all 🙏

https://codesandbox.io/s/flamboyant-nash-22lmm?file=/src/App.js

```jsx
const MySelect = props => {
  const [selected, setSelected] = useState(null);

  const handleChange = option => {
    if (typeof props.onChange === "function") {
      props.onChange(option.value); // <- pass result to Controller
    }
    setSelected(option);
  };

  useEffect(() => {
    if (typeof props.defaultVal !== "undefined" && props.defaultVal !== null) {
      const option = props.options.filter(opt => opt.value == props.defaultVal);
      if (option.length) {
        setSelected(option[0]);
        handleChange(option[0]); // <- trigger it when first time render
      }
    }
  }, []);

  return (
    <Select
      // {...props}  <- Prevent Controller using it directly
      isMulti={false}
      isSearchable={false}
      onChange={handleChange}
      options={props.options}
      value={selected}
      defaultValue={selected}
      placeholder={props.placeholder}
    />
  );
};
```

**Reply by bluebill1049** [reply]
here is a working version with V6: https://codesandbox.io/s/restless-rgb-e0q7y?file=/src/App.tsx

**Reply by Cow258** [reply]
I added
```jsx
<input type="button" value="Edit" onClick={()=>{
  setValue("select", "banana");
}}/>
```

Why dose it working?\
How RHF know `banana` is `{value:'banana',label:'Banana'}` and pass it to `Select`?

**Reply by bluebill1049** [reply]
did you take a look at the comments that I left in the CSB? `Controller` is alike data bridge and data can flow from `useForm` and it's methods.

**Reply by jalalmanafi** [reply]
> here is a working version with V6: https://codesandbox.io/s/restless-rgb-e0q7y?file=/src/App.tsx

onChange is not a function error

**Reply by Moshyfawn** [reply]
@jalalbmnf works fine for me. Check the versions of the packages you're using and try creating a reproduction, if something that you're doing is different

https://user-images.githubusercontent.com/16290753/152147099-25b04608-cc89-4f42-8b0e-6a0f4bb2544b.mp4

**Reply by jalalmanafi** [reply]
> @jalalbmnf works fine for me. Check the versions of the packages you're using and try creating a reproduction, if something that you're doing is different
> 
>  rhf-v6-onchange.mp4


Version : "react-hook-form": "^7.13.0",
Yeah , i am writing same code but still return same error. Here is my code



const partnerOptions = data?.map((item: IPartner) => {
    return {
      label: item?.name,
      value: item?.id,
    };
  });

 defaultValues: { partner: partnerOptions?.[0],
    },


 type SelectProps = React.ComponentProps<typeof Select>;

  const PartnerSelect = ({
    onChange,
    onBlur,
    value,
  }: Pick<SelectProps, "onChange" | "onBlur" | "value">) => {
    return (
      <Select
        options={partnerOptions}
        placeholder="Choose partner"
        styles={customerSelect}
        isSearchable={false}
        components={{
          IndicatorSeparator: () => null,
        }}
        {...{ onChange, onBlur, value }}
      />
    );
  };

 <Controller
              name="partner"
              control={control}
              render={({ onChange, onBlur, value }: any) => (
                <PartnerSelect
                  {...{
                    onChange: (item) => onChange(item),
                    onBlur,
                    value: partnerOptions?.find(
                      (item: any) => item?.value === value
                    ),
                  }}
                />
              )}
            />

**Reply by Moshyfawn** [reply]
The `Controller` API has changed since V6. In V7 there's a `field` property and inside it the `onChange` handler you want. Please, refer to [the Controller docs](https://react-hook-form.com/api/usecontroller/controller) for more info
```js
render({ field }) => (
  <input onChange={event => field.onChange(event.target.value)} />
)
```

**Reply by jalalmanafi** [reply]
> The `Controller` API has changed since V6. In V7 there's a `field` property and inside it the `onChange` handler you want. Please, refer to [the Controller docs](https://react-hook-form.com/api/usecontroller/controller) for more info
> 
> ```js
> render({ field }) => (
>   <input onChange={event => field.onChange(event.target.value)} />
> )
> ```

I dont't find anything about this issue at docs , already i looked before to docs

**Reply by Moshyfawn** [reply]
If you're using RHF **V7** _(you said `Version : "react-hook-form": "^7.13.0"`)_,
**instead of doing**
```js
render={({ onChange }) => (
  <PartnerSelect  onChange={item => onChange(item)} />
)}
```
you **need to do**
```js
render={({ field }) => (
  <PartnerSelect  onChange={item => field.onChange(item)} />
)}
```
Note the `field` property that the `onChange` handler is in in **V7**.

The example from **bluebill1049** is for the **V6** of RHF
> here is a working version with V6: https://codesandbox.io/s/restless-rgb-e0q7y?file=/src/App.tsx

**Reply by jalalmanafi** [reply]
FINALLY !!! WORKSS !!! THANK YOU !!!
